src/rdkafka_broker.c

```diff
@@ -3614,20 +3644,117 @@ rd_kafka_fetch_reply_handle (rd_kafka_broker_t *rkb,
+                                           compare the message offset with OFFSETS[ABORTED_TXN_OFFSETS(PID).IDX].
+                                           If it's greater than or equal to this value, then the message should be
+                                           ignored. If it's less than, then the message should be kept.
+
+                                        Note: A MessageSet comprises messages from at most one transaction, so the
+                                        logic in step 2 is done at the message set level.
+                                        */
+
+                                        for (k = 0 ; k < AbortedTxnCnt; k++) {
+                                                int64_t pid;
+                                                int64_t first_offset;
+                                                rd_kafka_buf_read_i64(rkbuf, &pid);
+                                                rd_kafka_buf_read_i64(rkbuf, &first_offset);
+
+                                                rd_kafka_aborted_txn_start_offsets_t node, *node_ptr;
+                                                node.pid = pid;
+
+                                                node_ptr = RD_AVL_FIND(aborted_txn_offsets, &node);
+                                                if (!node_ptr) {
+                                                        node_ptr = rd_malloc(sizeof(*node_ptr));
+                                                        node_ptr->pid = pid;
+                                                        node_ptr->offsets_idx = 0;
+                                                        rd_list_init(&node_ptr->offsets, 0, NULL);
+                                                        rd_list_prealloc_elems(&node_ptr->offsets,
+                                                                sizeof(int64_t), 
+                                                                AbortedTxnCnt, 0);
+                                                        RD_AVL_INSERT(aborted_txn_offsets, node_ptr, avl_node);
+                                                }
+
+                                                int64_t *v = rd_list_add(&node_ptr->offsets, NULL);
+                                                *v = first_offset;
+                                        }
+
+                                        /* sort each PID txn list */
+                                        rd_list_t queue;
+                                        rd_list_init(&queue, 0, NULL);
+                                        rd_list_add(&queue, aborted_txn_offsets->ravl_root);
+                                        while (!rd_list_empty(&queue)) {
+                                                rd_kafka_aborted_txn_start_offsets_t *el = rd_list_elem(&queue, 0);
+                                                rd_list_sort(&el->offsets, rd_kafka_offset_cmp);
+                                                rd_list_remove_elem(&queue, 0);
```

Comment:
using the list a stack would be better given the implementation of list, but I can't be bothered changing it and i think it will have no practical consequence.